[Bryce-3D] iP

Bryce-3D opened the pull request:
**DISCLAIMER**: This content was mostly shamelessly stolen from [the sample here](https://nus-cs2103-ay2425s2.github.io/website/schedule/week4/project.html#1-use-gfmd-in-the-pr-description)

# Homura
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

Homura frees your mind of having to remember things you need to do. It's
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is
1. download it from [here](https://github.com/Bryce-3D/ip) (?)
2. double-click it
3. add your tasks
4. let it manage your tasks for you 🧡 

And it is **FREE**! (until [enshittification](https://en.wikipedia.org/wiki/Enshittification) happens at least)

> [!WARNING]
> ⚠️ This will absolutely be buggy ⚠️ 

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
- [ ] Writing your code for you (Pls help me instead)

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

[a](a.com)

Have some random Python
```py
def gcd(a:int, b:int) -> int:
    a,b = abs(a),abs(b)
    if a<b:
        a,b = b,a
    while b != 0:
        a,b = b,a%b
    return a
```

Bryce-3D committed:
Refactor Homura.sc to Homura.SC

Bryce-3D committed:
Tidy up Homura.java and refactor functions to follow codestyle

Bryce-3D committed:
Rename variables in Homura.java from snake_case to camelCase

Bryce-3D committed:
Change functions in Todo.java from snake_case to camelCase

Bryce-3D committed:
Fix indentation levels to follow codestyle

Bryce-3D committed:
Make attributes public in Todo.java

Bryce-3D committed:
Make attributes private in Deadline and Event

Bryce-3D committed:
Homura.java: Make attributes private

Bryce-3D committed:
Homura.java: Change constants to public

Bryce-3D committed:
Add Storage.java to read and write files

Bryce-3D committed:
Rename Deadline.deadline to Deadline.by

Bryce-3D committed:
Add getters to Deadline and Event

Bryce-3D committed:
Implement methods to convert todo, etc, to a string for storage

Bryce-3D committed:
Create Playground.java so that I can test my code

Bryce-3D committed:
Implement split() in Storage.java

Bryce-3D committed:
Abstract out helper functions to HomuraUtils.java

Bryce-3D committed:
Implement Storage.from_str()

Bryce-3D committed:
Add Javadocs for Storage.from_str()

Bryce-3D committed:
Rename functions to follow codestyle

Bryce-3D committed:
Implement HomuraUtils.printArrList()

Bryce-3D committed:
Implement Storage.readTodos() and Storage.writeTodos()

Bryce-3D committed:
Add Javadocs to Homura.java

Bryce-3D committed:
Add newlines to Homura.java

Bryce-3D committed:
Rename Homura.tasks to Homura.todos

Bryce-3D committed:
Update Storage.readTodos() to return an empty list if no file

Bryce-3D committed:
Rearrange Storage.readTodos()

Bryce-3D committed:
Implement Homura.on() and Homura.off() with storage

Bryce-3D committed:
Fix Homura.on() and Homura.off() Javadocs

Bryce-3D committed:
Incorporate on() and off() into Homura.main()

Bryce-3D committed:
Fix bugs

Bryce-3D committed:
Merge branch 'branch-Level-7'

Bryce-3D committed:
Simplify HomuraUtils.split() implementation

Bryce-3D committed:
Implement toStorageStr() in Todo and subclasses

Bryce-3D committed:
Incorporate toStorageStr() into Storage.wrtiteTodos()

Bryce-3D committed:
Implement part of LocalDate update

Bryce-3D committed:
Implement fromStorageStr() in Todo and subclasses

Bryce-3D committed:
Change sta and end to LocalDate in Event.java

Bryce-3D committed:
Incorporate new fromStorageStr() functions into Storage.java

Bryce-3D committed:
Organise code in Todo and subclasses

Bryce-3D committed:
Organise Storage.java

Bryce-3D committed:
Add newlines to Homura.java

Bryce-3D committed:
Merge branch 'branch-Level-8'

Bryce-3D committed:
Add credits to ChatGPT in Deadline.java

Bryce-3D committed:
Implement Parser class

Bryce-3D committed:
Incorporate Parser.java into Homura.java

Bryce-3D committed:
Deprecate parseUserInput() from Todo and subclasses

Bryce-3D committed:
Modify runtest.sh to delete HomuraTodos.txt after each test

Bryce-3D committed:
Create TODO.txt

Bryce-3D committed:
Update Input and EXPECTED files

Bryce-3D committed:
Fix bug where Homura crashes with an empty HomuraTodos.txt

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Implement Ui.java

Bryce-3D committed:
Implement Ui.java (forgot to stage Homura.java)

Bryce-3D committed:
Incorporate Ui into Homura and depracate old ones from Homura

Bryce-3D committed:
Make runtest.bat delete HomuraTodos.txt and clean runtest.sh

Bryce-3D committed:
Implement TaskList.java

Bryce-3D committed:
Implement TaskList.java

Bryce-3D committed:
Change TaskList from all static to all non-static

Bryce-3D committed:
Add new constructor to TaskList.java

Bryce-3D committed:
Incorporate TaskList into Homura and Storage

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Merge branch 'branch-A-MoreOOP'

Bryce-3D committed:
Shove everything into a package named Homura

Bryce-3D committed:
Fix runtest.sh after it got bricked by packages

Bryce-3D committed:
Add ChatGPT's explanation into runtest.sh for future reference

Bryce-3D committed:
Fix runtest.bat after it got bricked by packages

Bryce-3D committed:
Merge branch 'branch-A-Packages'

Bryce-3D committed:
Merge branch 'add-gradle-support'

Bryce-3D committed:
Setup Gradle and update my TODO list

Bryce-3D committed:
Merge branch 'branch-A-Gradle'

Bryce-3D committed:
Fix build.gradle so that it can actually run

Bryce-3D committed:
Update build.gradle to handle JUnit Tests

Bryce-3D committed:
Initialise a test class for Todo

Bryce-3D committed:
Implement equals() in Todo and subclasses

Bryce-3D committed:
Implement basic parser tests

Bryce-3D committed:
Correct thrown exceptions in Parser

Bryce-3D committed:
Correct thrown exceptions in Parser

Bryce-3D committed:
Implement equals() in EmptyInputHomuraException

Bryce-3D committed:
Implement tests for Parser on empty inputs

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Flip expected and actual to the right spots in ParserTest

Bryce-3D committed:
Implement standard tests in TodoTest

Bryce-3D committed:
Tidy up ParserTest

Bryce-3D committed:
Add invalid input test to and tidy up TodoTest

Bryce-3D committed:
Merge branch 'branch-A-JUnit'

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Update build.gradle to be able to produce JAR files

Bryce-3D committed:
Merge branch 'branch-A-Jar'

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Add Javadocs to Error classes

Bryce-3D committed:
Make non-final attributes in error subclasses private

Bryce-3D committed:
Add Javadocs to Todo and subclasses

Bryce-3D committed:
Add Javadocs to 5 files

Bryce-3D committed:
Add Javadocs to Homura

Bryce-3D committed:
Implement contains() in Todo

Bryce-3D committed:
Implement findTodosWith() in TaskList

Bryce-3D committed:
Implement printTodos() and printFoundTodosFormatted() in Ui

Bryce-3D committed:
Implement cmdFind() and find functionality in Homura

Bryce-3D committed:
Merge branch 'branch-A-CodingStandard'

Bryce-3D committed:
Merge branch 'branch-A-Javadoc'

# Conflicts:
#	src/main/java/Homura/EmptyInputHomuraException.java

Bryce-3D committed:
Merge branch 'branch-Level-9'

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Refactor package Homura to homura and update build.gradle

Bryce-3D committed:
Make all lines under 80 chars long

Bryce-3D committed:
Update my TODO list

ArtemiszenN commented:
lgtm

Bryce-3D committed:
Add ability to use javaFX by adding in dependencies

Bryce-3D committed:
Initialize Launcher.java

Bryce-3D commented:
Test

Bryce-3D committed:
Add MadoHomu pfps and sauce in README.md

bipplane commented:
LGTM! though it could use more homura akemi ❤️

Bryce-3D committed:
Implement 2 functions in Ui.java for Homura

These new functions are needed for use in Homura.cmdFindJavafx()
since the current functionality only allows for directly printing rather than actually returning the String.

Bryce-3D committed:
Implement JavaFX versions of bot commands

Bryce-3D committed:
Implement Ui.todosFormatted()

Done in preparation for Homura.cmdJavaFX()

Bryce-3D committed:
Add a Javafx suffix to the new functions in Ui.java

Bryce-3D committed:
Implement Homura.cmdJavafx() for use in MainWindow.java

Bryce-3D committed:
Fix typo in Homura.java

Bryce-3D committed:
Implement DialogBox.java and DialogBox.fxml

Most of the code is based off the tutorial from
https://se-education.org/guides/tutorials/javaFx.html

Bryce-3D committed:
Implement MainWindow in Java and FXML

Most of the code is based off the tutorial from
https://se-education.org/guides/tutorials/javaFx.html

Bryce-3D committed:
Implement Launcher.java and Main.java

Most of the code is based off the tutorial from
https://se-education.org/guides/tutorials/javaFx.html

Bryce-3D committed:
Make Gradle load the GUI version instead

h-b-k-nishi commented on src/main/java/homura/TaskList.java:
```diff
@@ -0,0 +1,72 @@
+public class TaskList {
+    // Attributes + Getters and Setters ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    private ArrayList<Todo> todos;
+    public ArrayList<Todo> getTodos() { return todos; }
+    public void setTodos(ArrayList<Todo> todos) { this.todos = todos; }
```
Could you place the procedure inside the method on a separate line? I noticed the same issue in several other places as well.

Bryce-3D replied:
Hi! Yes that is indeed a coding standard violation. I tentatively plan to move it to a separate line towards the end of the iP. Thanks for pointing it out!

h-b-k-nishi commented on src/main/java/homura/Homura.java:
```diff
@@ -0,0 +1,356 @@
+    /**
+     * Turns the bot on.
+     */
+    public static void on() {
```
How about making the method name a verb and more specific?

Bryce-3D replied:
Hi! That is a good point, I'll rename it to `turnOn()` and `turnOff()`. Thanks!

h-b-k-nishi reviewed (commented):
LGTM. Just a few nits to fix ☺️

Bryce-3D committed:
Add ability for Homura to close on command "bye"

Bryce-3D committed:
Add setter for Homura.todos so it can be initialized outside

Bryce-3D committed:
Implement reading saved tasklist on startup

Bryce-3D committed:
Implement new goodbye message for the JavaFX version

Bryce-3D committed:
Make Homura (JavaFX) take an extra user message to exit

Bryce-3D committed:
Update the JavaFX version of Homura to use the new bye msg

Bryce-3D committed:
Add credits to ChatGPT for showing `Platform.exit();`

Bryce-3D committed:
Fix bug where closing Homura via chat causes errors in the log

Bryce-3D committed:
Implement cmdByeJavafx() and cmdListJavafx() in Homura.java

Bryce-3D committed:
Incorporate cmdByeJavafx() and cmdListJavafx() into the switch

Bryce-3D committed:
Implement new intro and bye msges for the JavaFX version

Bryce-3D committed:
Incorporate the new intro message in the JavaFX version

Bryce-3D committed:
Add missing newline

Bryce-3D committed:
Clear CLI formatting from 2 JavaFX functions in Ui

Bryce-3D committed:
Fix bug where find (JavaFX) prints out the result of list

Bryce-3D committed:
Remove commented out codes within Homura JavaFX cmd functions

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Implement new add/mark/unmark String functions for JavaFX in Todo

Bryce-3D committed:
Incorporate new JavaFX add/mark/unmark Strings into Homura

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Reorder cmd functions and add newlines in Homura

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Merge branch 'branch-Level-10'

Bryce-3D committed:
Modify build.gradle to generate a new jar file

Bryce-3D committed:
Rename on() and off() as suggested by a PR from h-b-k-nishi

Bryce-3D committed:
Add assertions to sanity check inputs in Parser.java

Bryce-3D committed:
Add asserts to sanity check inputs in JavaFX cmd functions

These asserts help to ensure that the JavaFX cmd functions do
not mistakenly get called when the input does not contain the
corresponding command

Bryce-3D committed:
Add an assert to sanity check the fromStorageStr() functions

These assertions help to ensure that the correct version of the
fromStorageStr() function is being called when parsing a storage
String.

Asserting that the first letter of the string corresponds to the
class of the fromStorageStr() being called helps to ensure this.

Bryce-3D committed:
Deprecate the CLI version by deprecating Homura.main()

Bryce-3D committed:
Deprecate cmd functions from the CLI version

Bryce-3D committed:
Recycle Javadocs from the original CLI cmd functions in Homura

Bryce-3D committed:
Deprecate CLI versions of add/mark/unmarkStr() in Todo

Bryce-3D committed:
Add explanatory comments in MainWindow.java

Bryce-3D committed:
Deprecate functions from the CLI version

Bryce-3D committed:
Deprecate attribute from the CLI version

Bryce-3D committed:
Add the Recycling Bin divider to DialogBox

Bryce-3D committed:
Add recycling bin divider to Main

Bryce-3D committed:
Add newline to end of MainWindow

Bryce-3D committed:
Clear out Playground

Bryce-3D committed:
Clear out all recycling bins (1)

Bryce-3D committed:
Remove unused attribute due to CLI deprecation

Bryce-3D committed:
Update cmdDeleteJavafx() formatting to one better for the GUI

Bryce-3D committed:
Merge branch 'master' into branch-A-Assertions

Bryce-3D committed:
Merge branch 'master' into branch-A-CodeQuality

Bryce-3D committed:
Remove unused attribute due to CLI deprecation

Bryce-3D committed:
Fill in empty catch blocks

Bryce-3D committed:
Merge pull request #1 from Bryce-3D/branch-A-CodeQuality

PR to Merge `branch-A-CodeQuality` to `master`

Bryce-3D committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/homura/Homura.java

Bryce-3D committed:
Merge pull request #2 from Bryce-3D/branch-A-Assertions

PR to Merge `branch-A-Assertions` to `master`

TasmiaH0508 commented on src/main/java/homura/Deadline.java:
```diff
@@ -0,0 +1,131 @@
+// and
+// https://www.w3schools.com/java/java_packages.asp
+package homura;
+
```
I see that you have tried to group all your classes into the same package. Perhaps you could further separate the classes to allow for better classification.

TasmiaH0508 commented on src/main/java/homura/Deadline.java:
```diff
@@ -0,0 +1,131 @@
+    @Override
+    public String toString() {
+        String ans = "[D]";
+        if (getIsDone()) {
```
```getIsDone()``` appears to be a boolean method. Is there any particular reason you chose to name this method this way instead of ```isDone()```, for instance?

Bryce-3D replied:
This one was named `getIsDone()` because it was meant as a getter for the `isDone` attribute. Naming the function `isDone()` again might not be ideal since then the attribute and function would have the same name. I'm not sure if there's a better way to name the getter. Thanks for pointing this out!

TasmiaH0508 commented on src/main/java/homura/Deadline.java:
```diff
@@ -0,0 +1,131 @@
+    @Override
+    public boolean equals(Object o) {
+        // Non-null Deadline
+        if (o == null) { return false; }
```
I think you may have accidentally forgotten about the requirement for Egyptian style braces... {Applied to other if statements within the coed block. }

Bryce-3D replied:
Hi! Yes, I forgot about this again, thanks for reminding me!

TasmiaH0508 commented on src/main/java/homura/EmptyInputHomuraException.java:
```diff
@@ -0,0 +1,59 @@
+package homura;
+
+public class EmptyInputHomuraException extends HomuraRuntimeException {
+    public static final String indent = "    ";
```
Perhaps the name of this variable should follow the Screaming Snake case, as stated by the java coding standards...

Bryce-3D replied:
I completely missed that, thanks for pointing it out!

TasmiaH0508 commented on src/main/java/homura/Event.java:
```diff
@@ -0,0 +1,137 @@
+    // Attributes + Getters and Setters ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    // Todo.str  description
+    // Todo.bool isDone
+    private LocalDate sta;
```
Perhaps you could rename the ```sta``` variable to ```start```, so as to prevent other readers from being confused about what ```sta``` is supposed to represent, later in the code.

TasmiaH0508 commented on src/main/java/homura/Homura.java:
```diff
@@ -0,0 +1,181 @@
+        return Ui.byeMsgJavafx();
+    }
+    
+    public static String cmdJavafx(String inp) {
```
Perhaps instead of ```cmdJavafx```, ```commandJavafx``` would be a better name for this method, following the Java coding standard

TasmiaH0508 reviewed (commented):
Hello, I was reviewing your code for code quality. 

I left some comments related to some violations on the coding standard. Apart from that, good job on the progress you've made so far.

Bryce-3D committed:
Fix some codestyle issues

h-b-k-nishi commented on src/main/java/homura/HomuraUtils.java:
```diff
@@ -0,0 +1,128 @@
+// https://github.com/weiseng18/ip/tree/master/src/main/java/panorama
+// and
+// https://www.w3schools.com/java/java_packages.asp
+package homura;
```
It's amazing you put all the classes into the package!

h-b-k-nishi commented on src/main/java/homura/Homura.java:
```diff
@@ -0,0 +1,356 @@
+
+
+
+// RECYCLING BIN ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
How about moving this "recycling bin" to another file, and keeping this original file tidy? Although I am not sure if it is a coding standard violation.

Bryce-3D committed:
Fix potential bug of " bye" not hitting the edited if clause

Bryce-3D committed:
Fix bug of the assert in Todo.fromStorageStr() failing a test

Bryce-3D committed:
Merge branch 'master' into branch-A-CI

Bryce-3D committed:
Create first version of GitHub CI

Bryce-3D committed:
Create empty version of gradle.yml

Bryce-3D committed:
Merge branch 'master' into branch-A-CI

# Conflicts:
#	.github/workflows/gradle.yml

Bryce-3D committed:
Update to run CI on pushes to master and pull requests

Bryce-3D committed:
Test pushing to master

Bryce-3D committed:
Remove test comment

Bryce-3D committed:
Fix branch name from main to master in GitHub CI

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Fix minor codestyle issue

Bryce-3D committed:
Merge branch 'master' into branch-BCD-Extension

Bryce-3D committed:
Copy set() functionality from ArrayList to TaskList

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Implement Todo.edit() in preparation for Homura.cmdEditJavafx()

Bryce-3D committed:
Override edit() in Todo's subclasses for Homura.cmdEditJavafx()

Bryce-3D committed:
Implement first version of Homura.cmdEditJavafx()

Currently has a bug where it cannot handle descriptions
with whitespaces in it

Bryce-3D committed:
Fix Homura.cmdEditJavafx() to handle whitespaces in des

Bryce-3D committed:
Clean up code in Homura.cmdEditJavafx()

Bryce-3D committed:
Merge branch 'branch-BCD-Extension'

Bryce-3D committed:
Remove deprecated functions from the CLI version

Bryce-3D committed:
Add Javadocs to Ui.java

Bryce-3D committed:
Use the DRY principle in Ui.java

Bryce-3D committed:
Clear out code in Playground.java

Bryce-3D committed:
Clean out comments in Homura.java

Bryce-3D committed:
Add Javadocs to cmd functions in Homura.java

Bryce-3D committed:
Trim function names in Homura.java by removing Javafx suffixes

Bryce-3D committed:
Remove Javafx suffix from Ui.java functions

Bryce-3D committed:
Remove unneeded comments from Ui.java

Bryce-3D committed:
Remove unused attribute in Ui.java

Bryce-3D committed:
Merge branch 'master' into branch-MoveAddMarkUnmarkOutOfTodo

Bryce-3D committed:
Implement functions in Ui.java to migrate from Todo.java

Bryce-3D committed:
Incorporate new Ui add/mark/unmark functions into Homura

Bryce-3D committed:
Add Javadocs to new Ui functions

Bryce-3D committed:
Deprecate add/mark/unmark str functions in Todo.java

Bryce-3D committed:
Merge branch 'branch-2025-02-11_MoveAddMarkUnmarkOutOfTodo'

Bryce-3D committed:
Delete deprecated tests for the CLI version

Bryce-3D committed:
Remove unnecessary ChatGPT citations as instructed in Issue 186

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Remove unnecessary citation for how to use Java packages

Bryce-3D committed:
Fix temporarily bug of Storage.readTodos() causing newline bugs

Bryce-3D committed:
Implement a new utility function to split along lines

Bryce-3D committed:
Incorporate the new splitLine() function into Storage.java

Bryce-3D committed:
Delete the now unnecessary strips in Deadline and Event

Bryce-3D committed:
Merge branch 'branch-2025-02-11_FixStorageReadTodosNewlineBug'

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Organize order of commands in Homura.java

Bryce-3D committed:
Remove probably unnecessary comments in Homura.java

Bryce-3D committed:
Delete unused field in Todo and clear out Playground

Bryce-3D committed:
Delete recycling bins (the bins themselves, not the contents)

Bryce-3D committed:
Fix wrongly called function in Homura.cmdUnmark()

Bryce-3D committed:
Remove unnecessary comments in GUI files

Bryce-3D committed:
Allow Homura to respond to invalid cmds and strip() the inp

Bryce-3D committed:
Add try blocks to todo,deadline,event and assert to todo,deadline

Bryce-3D committed:
Add detail to Homura's responses for erroneous todo,etc creation

Bryce-3D committed:
Modify exception handling in commands modifying todos

Bryce-3D committed:
Merge branch 'branch-A-MoreErrorHandling'

Bryce-3D committed:
Add Javadocs to Todo and subclasses

Bryce-3D committed:
Add Javadocs to remaining public functions

Bryce-3D committed:
Move class Javadocs to right before the class

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Record ChatGPT queries

Bryce-3D committed:
Add representative screenshot of Homura

Bryce-3D committed:
Fix typo in output of list and find commands

Bryce-3D committed:
Change bye message of Homura

Bryce-3D committed:
Update the user guide with instructions on how to use Homura

Bryce-3D committed:
Make Homura not include nulls when reading the storage file

Bryce-3D committed:
Add section on manually deleting the todo list

Bryce-3D committed:
Make the find cmd case sensitive and fix bugs with it

Bryce-3D committed:
Add user guide for find cmd and fix typos

Bryce-3D committed:
Fix typo of commented out text not being commented out

Bryce-3D committed:
Fix code quality as in https://github.com/Bryce-3D/ip/issues/3

Bryce-3D committed:
Make Homura provide the link to the user guide on invalid cmds

Bryce-3D committed:
Change ✓ to X to prevent buggy character encoding

Bryce-3D committed:
Add setup and execution to the user guide

Bryce-3D committed:
Update build.gradle

Bryce-3D committed:
Update my TODO list

Bryce-3D committed:
Update Ui.png to reflect using X for todos marked as done
